### PR TITLE
Nydusify: support "view" subcommand to view Nydus image

### PR DESCRIPTION
**Pull request by YushuoEdge**
Nydusify provides a fsViewer to view the file system in the Nydus image.
The fsViewer will pull image's bootstrap according to its url, and it
will mount the file system to pointed path.

Users can view the file system then. For example, they can ran
`$ trivy fs mount-path`
to scan the whole image.

Signed-off-by: YushuoEdge <y-shuo@sjtu.edu.cn>

### CODE REVIEW

**Comment by yqleng1987** [comment]
@YushuoEdge , a new test job has been submitted. Please wait in patience. The test job url: [https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445](https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445)

**Comment by yqleng1987** [comment]
@YushuoEdge , The CI test is completed, please check result:
<table>
    <tr>
        <th colspan="2">Test Case</th><th>Test Result</th>
    </tr>
    <tr>
        <td colspan="2">merge-target-branch</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">build-docker-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-ctr-remote</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus-snapshotter</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">start-nydus-snapshotter-config-containerd</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">run-container-with-nydus-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13445>:white_check_mark: SUCCESS</a></td>
    </tr>
</table>

Congratulations, your test job passed!

**Comment by yqleng1987** [comment]
@YushuoEdge , the code has been updated, so a new test job has been submitted. Please wait in patience. The test job url: [https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446](https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446)

**Comment by yqleng1987** [comment]
@YushuoEdge , The CI test is completed, please check result:
<table>
    <tr>
        <th colspan="2">Test Case</th><th>Test Result</th>
    </tr>
    <tr>
        <td colspan="2">merge-target-branch</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">build-docker-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-ctr-remote</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus-snapshotter</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">start-nydus-snapshotter-config-containerd</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">run-container-with-nydus-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/13446>:white_check_mark: SUCCESS</a></td>
    </tr>
</table>

Congratulations, your test job passed!

File: contrib/nydusify/cmd/nydusify.go [hunk 1]
```diff
@@ -397,6 +398,58 @@ func main() {
...
+		{
+			Name:  "view",
+			Usage: "View the file system in nydus image",
+			Flags: []cli.Flag{
+				&cli.StringFlag{Name: "target", Required: true, Usage: "Target (Nydus) image reference", EnvVars: []string{"TARGET"}},
+				&cli.BoolFlag{Name: "target-insecure", Required: false, Usage: "Allow http/insecure target registry communication", EnvVars: []string{"TARGET_INSECURE"}},
+
+				&cli.StringFlag{Name: "mount-path", Value: "./image-fs", Usage: "The Path for the file system to mount on", EnvVars: []string{"MOUNT_PATH"}},
+				&cli.StringFlag{Name: "platform", Value: "linux/" + runtime.GOARCH, Usage: "Let nydusify choose image of specified platform from manifest index. Possible value is `amd64` or `arm64`"},
+				&cli.StringFlag{Name: "work-dir", Value: "./tmp", Usage: "Work directory path for image check, will be cleaned before checking", EnvVars: []string{"WORK_DIR"}},
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:410]: `will be cleaned before viewing`

File: contrib/nydusify/cmd/nydusify.go [hunk 2]
```diff
@@ -397,6 +398,58 @@ func main() {
...
+				&cli.StringFlag{Name: "nydusd", Value: "nydusd", Usage: "The nydusd binary path, if unset, search in PATH environment", EnvVars: []string{"NYDUSD"}},
+				&cli.StringFlag{Name: "backend-type", Value: "", Usage: "Specify Nydus blob storage backend type, will check file data in Nydus image if specified", EnvVars: []string{"BACKEND_TYPE"}},
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:412]: `will allow viewing file data in Nydus image if specified`

File: contrib/nydusify/pkg/checker/tool/nydusd.go [hunk 3]
```diff
@@ -203,11 +205,14 @@ func (nydusd *Nydusd) Mount() error {
 	return nil
 }
 
-func (nydusd *Nydusd) Umount() error {
+func (nydusd *Nydusd) Umount(mountFlag bool) error {
```
Comment (imeoer) [inline contrib/nydusify/pkg/checker/tool/nydusd.go:208]: Better to use `silent bool`.

File: contrib/nydusify/pkg/viewer/viewer.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	return fsViewer, nil
+}
+
+// Pull Bootstrap, includes nydus_manifest.json and nydus_config.json
+func (fsViewer *FsViewer) PullBootstrap(ctx context.Context, targetParsed *parser.Parsed) error {
+	if err := os.RemoveAll(fsViewer.WorkDir); err != nil {
+		return errors.Wrap(err, "clean up work directory error")
+	}
+
+	if err := os.MkdirAll(filepath.Join(fsViewer.WorkDir, "fs"), 0755); err != nil {
```
Comment (imeoer) [inline contrib/nydusify/pkg/viewer/viewer.go:100]: Using `0750` is safer.

File: contrib/nydusify/pkg/viewer/viewer.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	return nil
+}
+
+// Mount nydus image.
+func (fsViewer *FsViewer) MountImage() error {
+	logrus.Infof("Mounting Nydus image to %s", fsViewer.NydusdConfig.MountPath)
+
+	if err := os.MkdirAll(fsViewer.NydusdConfig.BlobCacheDir, 0755); err != nil {
```
Comment (imeoer) [inline contrib/nydusify/pkg/viewer/viewer.go:138]: Ditto.

File: contrib/nydusify/pkg/viewer/viewer.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		return errors.Wrap(err, "create blob cache directory for Nydusd error")
+	}
+
+	if err := os.MkdirAll(fsViewer.NydusdConfig.MountPath, 0755); err != nil {
```
Comment (imeoer) [inline contrib/nydusify/pkg/viewer/viewer.go:142]: Ditto.

File: contrib/nydusify/pkg/viewer/viewer.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	// Block current goroutine in order to umount the file system and clean up workdir
+	sigs := make(chan os.Signal, 1)
+	done := make(chan bool, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		sig := <-sigs
+		fmt.Println("Received Signal: ", sig)
```
Comment (imeoer) [inline contrib/nydusify/pkg/viewer/viewer.go:184]: `logrus.Infof("Received Signal: %s", sig)` ?

File: contrib/nydusify/pkg/viewer/viewer.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+		done <- true
+	}()
+
+	fmt.Println("Please send signal SIGINT/SIGTERM to umount the file system")
```
Comment (imeoer) [inline contrib/nydusify/pkg/viewer/viewer.go:188]: Ditto, use `logrus.Info`.

File: contrib/nydusify/cmd/nydusify.go [hunk 9]
```diff
@@ -397,6 +398,58 @@ func main() {
 				return checker.Check(context.Background())
 			},
 		},
+		{
+			Name:  "view",
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:402]: Maybe we can add an alias `Aliases: []string{"mount"},`.

File: contrib/nydusify/cmd/nydusify.go (hunk 1 above)
Comment (YushuoEdge) [reply contrib/nydusify/cmd/nydusify.go:410]: yes，maybe we should give a hint?

File: contrib/nydusify/cmd/nydusify.go (hunk 2 above)
Comment (YushuoEdge) [reply contrib/nydusify/cmd/nydusify.go:412]: Should we specify a specific backend type here?

File: contrib/nydusify/pkg/viewer/viewer.go (hunk 4 above)
Comment (YushuoEdge) [reply contrib/nydusify/pkg/viewer/viewer.go:100]: OK!

File: contrib/nydusify/pkg/viewer/viewer.go (hunk 8 above)
Comment (YushuoEdge) [reply contrib/nydusify/pkg/viewer/viewer.go:188]: OK!

File: contrib/nydusify/cmd/nydusify.go (hunk 9 above)
Comment (YushuoEdge) [reply contrib/nydusify/cmd/nydusify.go:402]: OK!

**Comment by yqleng1987** [comment]
@YushuoEdge , the code has been updated, so a new test job has been submitted. Please wait in patience. The test job url: [https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298](https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298)

**Comment by yqleng1987** [comment]
@YushuoEdge , The CI test is completed, please check result:
<table>
    <tr>
        <th colspan="2">Test Case</th><th>Test Result</th>
    </tr>
    <tr>
        <td colspan="2">merge-target-branch</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">build-docker-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-ctr-remote</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus-snapshotter</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">start-nydus-snapshotter-config-containerd</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">run-container-with-nydus-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15298>:white_check_mark: SUCCESS</a></td>
    </tr>
</table>

Congratulations, your test job passed!

**Comment by yqleng1987** [comment]
@YushuoEdge , The CI test is completed, please check result:
<table>
    <tr>
        <th colspan="2">Test Case</th><th>Test Result</th>
    </tr>
    <tr>
        <td colspan="2">merge-target-branch</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">build-docker-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-ctr-remote</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">compile-nydus-snapshotter</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">start-nydus-snapshotter-config-containerd</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
    <tr>
        <td colspan="2">run-container-with-nydus-image</td><td><a href=https://tone.openanolis.cn/ws/nrh4nnio/test_result/15299>:white_check_mark: SUCCESS</a></td>
    </tr>
</table>

Congratulations, your test job passed!

File: contrib/nydusify/pkg/viewer/viewer.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+	MountPath     string
+	NydusdPath    string
+	BackendType   string
+	BackendConfig string
+	ExpectedArch  string
+	FsVersion     string
+}
+
+// fsViewer provides complete view of file system in nydus image
+type FsViewer struct {
```
Comment (imeoer) [inline contrib/nydusify/pkg/viewer/viewer.go:44]: `FsViewer provides complete view of file system in nydus image`

**Review by imeoer — APPROVED** [review]
LGTM!

**Pull request merged**